src/storage-operators/src/oneshot_source.rs

```diff
@@ -0,0 +1,892 @@
+            .open(persist_location)
+            .await
+            .expect("failed to open Persist client");
+        let persist_diagnostics = Diagnostics {
+            shard_name: collection_id.to_string(),
+            handle_purpose: "CopyFrom::stage_batches".to_string(),
+        };
+        let write_handle = persist_client
+            .open_writer::<SourceData, (), mz_repr::Timestamp, Diff>(
+                shard_id,
+                Arc::new(collection_desc),
+                Arc::new(UnitSchema),
+                persist_diagnostics,
+            )
+            .await
+            .expect("could not open Persist shard");
+
+        // TODO(parkmcar): Use different timestamps here.
+        let lower = mz_repr::Timestamp::MIN;
+        let upper = Antichain::from_elem(lower.step_forward());
+
+        let mut batch_builder = write_handle.builder(Antichain::from_elem(lower));
+        let mut caps = CapabilitySet::new();
+
+        while let Some(event) = rows_handle.next().await {
+            let AsyncEvent::Data(cap, row_batch) = event else {
+                continue;
+            };
+
+            // Note: In the current implementation of `COPY FROM` this set should
+            // only ever contain one capability.
+            caps.insert(cap);
+
+            // Pull Rows off our stream and stage them into a Batch.
+            for maybe_row in row_batch {
+                match maybe_row {
+                    // Happy path, add the Row to our batch!
+                    Ok(row) => {
+                        let data = SourceData(Ok(row));
+                        batch_builder
+                            .add(&data, &(), &lower, &1)
+                            .await
+                            .expect("failed to add Row to batch");
+                    }
+                    // Sad path, something upstream hit an error.
+                    Err(err) => {
+                        // Clean up our in-progress batch so we don't leak data.
+                        let batch = batch_builder
+                            .finish(upper)
+                            .await
+                            .expect("failed to cleanup batch");
+                        batch.delete().await;
+
+                        // Pass on the error.
+                        let send_cap = caps.first().expect("didn't get any capabilities?");
+                        proto_batch_handle.give(send_cap, Err(err).context("stage batches"));
+                        return;
+                    }
+                }
+            }
+        }
+
+        let batch = batch_builder
+            .finish(upper)
+            .await
+            .expect("failed to create Batch");
+
+        // Turn out Batch into a ProtoBatch that will later be linked in to
+        // the shard.
+        //
+        // Note: By turning this into a ProtoBatch, the onus is now on us to
+        // cleanup the Batch if it's never linked into the shard.
```

Comment:
On who was the onus before? Does persist automatically garbage collect unlinked batches?

Reply:
If Rust had `async Drop` then I believe we would automatically clean up unlinked `Batch`es, [today we just warn though](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/batch.rs#L95-L109). Really the point of this comment was to echo what [the comment on into_transmittable_batch()](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/batch.rs#L203-L206) says.